### PR TITLE
Get PunchItems by checklist guid

### PR DESCRIPTION
**Pull request by Hatleskog**

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using Equinor.ProCoSys.Completion.ForeignApi.MainApi.CheckList;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetCheckListsByPunchItemGuid;
+
+public class GetCheckListsByPIGuidQuery(Guid punchItemGuid) : IRequest<Result<ChecklistsByPunchGuidInstance>>
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs:8]: GetCheckListsByPunchItemGuidQuery

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQueryHandler.cs [hunk 2]
```diff
@@ -5,14 +5,14 @@
 using MediatR;
 using ServiceResult;
 
-namespace Equinor.ProCoSys.Completion.Command.PunchItemCommands.GetCheckListsByPunchItemGuid;
+namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetCheckListsByPunchItemGuid;
 
-public class GetCheckListsByPIGuidCommandHandler : IRequestHandler<GetCheckListsByPIGuidCommand, Result<ChecklistsByPunchGuidInstance>>
+public class GetCheckListsByPIGuidQueryHandler : IRequestHandler<GetCheckListsByPIGuidQuery, Result<ChecklistsByPunchGuidInstance>>
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQueryHandler.cs:10]: GetCheckListsByPunchItemGuidQueryHandler

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs (hunk 1 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs:8]: Inherit IIsPunchItemQuery. 
In namespace Equinor.ProCoSys.Completion.WebApi.Tests.Authorizations.IsPunchItemQueryTests, add test AccessValidatorForGetCheckListsByPunchItemGuidQueryTests

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemDetailsDto.cs [hunk 3]
```diff
@@ -40,4 +40,5 @@ public sealed record PunchItemDetailsDto(
     DocumentDto? Document,
     SWCRDto? SWCR,
     int AttachmentCount,
-    string RowVersion);
+    string RowVersion,
+    Guid? CheckListGuid);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemDetailsDto.cs:44]: make this not nullable. Move the arg as second arg, between Guid and ProjectName.

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs (hunk 1 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQuery.cs:8]: I'm unsure if it's correct that this Query should return the exact same dto as main api has returned, or if we should make an internal dto to map to. Ask Tord or Sindre for second opinion

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQueryHandler.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPIGuidQueryHandler.cs:10]: Add unit test, test that the handler return the dto returned from _checkListApiService.GetByPunchItemGuidAsync

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItem/GetPunchItemQueryHandler.cs [hunk 4]
```diff
@@ -1,146 +1,20 @@
...
 
 namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetPunchItem;
 
 public class GetPunchItemQueryHandler : IRequestHandler<GetPunchItemQuery, Result<PunchItemDetailsDto>>
 {
-    private readonly IReadOnlyContext _context;
+    private readonly IPunchItemService _punchItemService;
 
-    public GetPunchItemQueryHandler(IReadOnlyContext context) => _context = context;
+    public GetPunchItemQueryHandler(IPunchItemService punchItemService) => _punchItemService = punchItemService;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItem/GetPunchItemQueryHandler.cs:13]: Add unit test, test that Hander return the dto returned from _punchItemService.GetByGuid

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Equinor.ProCoSys.Completion.Query.PunchItemServices;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetPunchItemsByCheckList;
+
+public class GetPunchItemsByCheckListGuidQueryHandler : IRequestHandler<GetPunchItemsByCheckListGuidQuery, Result<List<PunchItemDetailsDto>>>
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandler.cs:10]: Add unit test, test that Hander return the dtos returned from _punchItemService.GetByCheckListGuid

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        RuleLevelCascadeMode = CascadeMode.Stop;
+        ClassLevelCascadeMode = CascadeMode.Stop;
+
+        RuleFor(query => query)
+            .Must((query, cancellationToken) => IsValidGuid(query.CheckListGuid))
+            .WithMessage(query => $"CheckListGuid is not valid! Guid={query.CheckListGuid}")
+            .WithState(_ => new ArgumentException());
+    }
+
+    private static bool IsValidGuid(Guid checkListGuid) => (Guid.Empty != checkListGuid);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs:19]: This is an input validator, which we don't need.
What we do need, is to check if the given checklist guid is an existing guid. See checkListValidator.ExistsAsync in CreatePunchItemCommandValidator in Equinor.ProCoSys.Completion.Command.PunchItemCommands.CreatePunchItem ns

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/IPunchItemService.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Equinor.ProCoSys.Completion.Query.PunchItemServices;
+
+public interface IPunchItemService
+{
+    Task<PunchItemDetailsDto> GetByGuid(Guid guid, CancellationToken cancellationToken);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/IPunchItemService.cs:10]: punchItemGuid

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/IPunchItemService.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    Task<List<PunchItemDetailsDto>> GetByCheckListGuid(Guid guid, CancellationToken cancellationToken);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/IPunchItemService.cs:11]: punchItemGuid

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+namespace Equinor.ProCoSys.Completion.Query.PunchItemServices;
+
+public class PunchItemService(IReadOnlyContext context) : IPunchItemService
+{
+    public async Task<PunchItemDetailsDto> GetByGuid(Guid guid, CancellationToken cancellationToken)
+    {
+        var whereClause = new Func<IQueryable<PunchItem>, IQueryable<PunchItem>>(query =>
+            query.Where(pi => pi.Guid == guid)
+        );
+        var punchItem = (await GetPunchItems(whereClause, cancellationToken)).FirstOrDefault()
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:27]: Use .First(), and remove the throw exception. No need for checking existance here, the GetPunchItemQueryValidator take car of that

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        return MapPunchToDto(punchItem, attCount);
+    }
+
+    public async Task<List<PunchItemDetailsDto>> GetByCheckListGuid(Guid guid, CancellationToken cancellationToken)
+    {
+        var whereClause = new Func<IQueryable<PunchItem>, IQueryable<PunchItem>>(query =>
+            query.Where(pi => pi.CheckListGuid == guid)
+        );
+        var punchItems = await GetPunchItems(whereClause, cancellationToken);
+        var punchItemGuids = punchItems.Select(y => y.Guid);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:41]: p => p.Guid

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+public class PunchItemService(IReadOnlyContext context) : IPunchItemService
+{
+    public async Task<PunchItemDetailsDto> GetByGuid(Guid guid, CancellationToken cancellationToken)
+    {
+        var whereClause = new Func<IQueryable<PunchItem>, IQueryable<PunchItem>>(query =>
+            query.Where(pi => pi.Guid == guid)
+        );
+        var punchItem = (await GetPunchItems(whereClause, cancellationToken)).FirstOrDefault() 
+                        ?? throw new EntityNotFoundException<PunchItem>(guid);
+        var attCount = await context.QuerySet<Attachment>()
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:29]: Add unit test, checking for attachment count > 0

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        var punchItems = await GetPunchItems(whereClause, cancellationToken);
+        var punchItemGuids = punchItems.Select(y => y.Guid);
+
+        var attachmentCounts = await context.QuerySet<Attachment>()
+            .Where(a => punchItemGuids.Contains(a.ParentGuid))
+            .GroupBy(a => a.ParentGuid)
+            .Select(g => new { PunchItemGuid = g.Key, AttachmentCount = g.Count() })
+            .ToListAsync(cancellationToken);
+
+        return punchItems.Select(pi => MapPunchToDto(pi,
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:49]: Add unit test, checking for both attachment count = 0 and attachment count > 0

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:49]: is this refactor better to read?
image

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/IAccessChecker.cs [hunk 13]
```diff
@@ -3,8 +3,9 @@
 
 namespace Equinor.ProCoSys.Completion.WebApi.Authorizations;
 
-public interface IContentAccessChecker
+public interface IAccessChecker
 {
-    Task<bool> HasCurrentUserAccessToCheckListAsync(Guid checkListGuid);
+    Task<bool> HasCurrentUserWriteAccessToCheckListAsync(Guid checkListGuid);
+    Task<bool> HasCurrentUserReadAccessToCheckListAsync(Guid checkListGuid);
     Task<bool> HasCurrentUserAccessToCheckListOwningPunchItemAsync(Guid punchItemGuid);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Authorizations/IAccessChecker.cs:10]: HasCurrentUserWriteAccessToCheckListOwningPunchItemAsync

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessValidator.cs [hunk 14]
```diff
@@ -100,6 +100,16 @@ public class AccessValidator : IAccessValidator
             }
         }
 
+        if (request is IIsCheckListQuery getPunchItemsByCheckListGuidQuery)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessValidator.cs:103]: checkListQuery

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessValidator.cs [hunk 15]
```diff
@@ -100,6 +100,16 @@ public class AccessValidator : IAccessValidator
...
+        {
+            if (!await _accessChecker.HasCurrentUserReadAccessToCheckListAsync(getPunchItemsByCheckListGuidQuery.CheckListGuid))
+            {
+                _logger.LogWarning("Current user {UserOid} doesn't have read access to checkList {CheckListGuid}",
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessValidator.cs:107]: go through other warnings and add phrase "read access" when request is a Query and add phrase "write access" when request is a Command

File: src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessChecker.cs [hunk 16]
```diff
@@ -49,6 +53,14 @@ public async Task<bool> HasCurrentUserAccessToCheckListOwningPunchItemAsync(Guid
         return await HasCurrentUserExplicitAccessToContent(checkListGuid.Value);
     }
 
+    public async Task<bool> HasCurrentUserReadAccessToCheckListAsync(Guid checkListGuid)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Authorizations/AccessChecker.cs:56]: Add unit tests, both when GetCheckListAsync return null and return a checklist

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidatorTests.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+        // Act
+        var result = await _dut.ValidateAsync(query);
+
+        // Assert
+        Assert.IsTrue(result.IsValid);
+    }
+
+    [TestMethod]
+    public async Task Validate_ShouldFail_When_Guid_Is_Empty()
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidatorTests.cs:30]: refactor test, test ShouldFail when checklist don't exists

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PunchItems/PunchItemsController.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Controllers/PunchItems/PunchItemsController.cs:1]: Add integration tests, both negative and positive tests for new endpoints.
For the positive test, you need to mock what the main api returns

The  negative tests are most important:  test "not found", "access denied" and so on

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

File: src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/ProCoSys4CheckListValidator.cs [hunk 18]
```diff
@@ -12,7 +13,7 @@ public class ProCoSys4CheckListValidator : ICheckListValidator
 
     public ProCoSys4CheckListValidator(ICheckListCache checkListCache) => _checkListCache = checkListCache;
 
-    public async Task<bool> ExistsAsync(Guid checkListGuid)
+    public async Task<bool> ExistsAsync(Guid checkListGuid, CancellationToken cancellationToken)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.ForeignApi/MainApi/CheckList/ProCoSys4CheckListValidator.cs:16]: why take in cancellationToken when you don't pass it on?

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs [hunk 19]
```diff
@@ -1,20 +1,26 @@
 ﻿using System;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs:1]: remove

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs [hunk 20]
```diff
@@ -1,20 +1,26 @@
...
         RuleFor(query => query)
-            .Must((query, cancellationToken) => IsValidGuid(query.CheckListGuid))
-            .WithMessage(query => $"CheckListGuid is not valid! Guid={query.CheckListGuid}")
-            .WithState(_ => new ArgumentException());
+            
+            // validate given CheckList
+            .MustAsync(BeAnExistingCheckListAsync)
+            .WithMessage(query => $"CheckList does not exist with Guid '{query.CheckListGuid}'")
+            .WithState(_ => new EntityNotFoundException());
+        async Task<bool> BeAnExistingCheckListAsync(GetPunchItemsByCheckListGuidQuery query, CancellationToken cancellationToken)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryValidator.cs:23]: add blnak line

File: src/Equinor.ProCoSys.Completion.WebApi/Controllers/PunchItems/PunchItemsController.cs [hunk 21]
```diff
@@ -123,7 +123,7 @@ public PunchItemsController(IMediator mediator, ILogger<PunchItemsController> lo
     /// <response code="404">Project not found</response>
     [AuthorizeAny(Permissions.PUNCHITEM_READ, Permissions.APPLICATION_TESTER)]
     [HttpGet("CheckList/{checkListGuid}")]
-    public async Task<ActionResult<IEnumerable<PunchItemDetailsDto>>> GetPunchItemsByCheckListGuid(
+    public async Task<ActionResult<IReadOnlyCollection<PunchItemDetailsDto>>> GetPunchItemsByCheckListGuid(
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/Controllers/PunchItems/PunchItemsController.cs:126]: curious: why is IReadOnlyCollection better than IEnumerable here?

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandler.cs [hunk 22]
```diff
@@ -21,7 +21,7 @@ ICheckListApiService checkListApiService
         _checkListApiService = checkListApiService;
     }
 
-    public async Task<Result<ChecklistsByPunchGuidInstance>> Handle(GetCheckListsByPIGuidCommand request,
+    public async Task<Result<ChecklistsByPunchGuidInstance>> Handle(GetCheckListsByPunchItemGuidQuery request,
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandler.cs:24]: at line 28, remove the ! behind res

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandlerTests.cs [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+        // Act
+        var result = await _dut.Handle(_query, default);
+        var data = result.Data;
+
+        // Assert
+        Assert.IsNotNull(result);
+        Assert.AreEqual(ResultType.Ok, result.ResultType);
+        Assert.IsNotNull(data.PunchItemCheckList);
+        Assert.AreEqual(_checkListGuid, data.PunchItemCheckList.ProCoSysGuid);
+        Assert.IsTrue(_checkListGuid == data.PunchItemCheckList.ProCoSysGuid);
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandlerTests.cs:53]: remove (same test as line above)

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandlerTests.cs [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+        Assert.IsTrue(1 == data.CheckLists.Count);
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetCheckListsByPunchItemGuid/GetCheckListsByPunchItemGuidQueryHandlerTests.cs:54]: Assert.AreEqual(1, data.CheckLists.Count); AreEqual is preffered since the output is better when test fail

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs [hunk 25]
```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs:4]: remove

File: src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandler.cs [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+using ServiceResult;
+
+namespace Equinor.ProCoSys.Completion.Query.PunchItemQueries.GetPunchItemsByCheckList;
+
+public class GetPunchItemsByCheckListGuidQueryHandler : IRequestHandler<GetPunchItemsByCheckListGuidQuery, Result<IEnumerable<PunchItemDetailsDto>>>
+{
+    private readonly IPunchItemService _punchItemService;
+
+    public GetPunchItemsByCheckListGuidQueryHandler(IPunchItemService punchItemService) => _punchItemService = punchItemService;
+    public async Task<Result<IEnumerable<PunchItemDetailsDto>>> Handle(GetPunchItemsByCheckListGuidQuery request, CancellationToken cancellationToken)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandler.cs:15]: add blank line

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+    public void Setup_OkState()
+    {
+        _raisedByOrg = new LibraryItemDto(Guid.NewGuid(), Code, "raisedBy", LibraryType.COMPLETION_ORGANIZATION.ToString()); 
+        _clearingByOrg = new LibraryItemDto(Guid.NewGuid(), Code, "clearingBy", LibraryType.COMPLETION_ORGANIZATION.ToString());
+
+        _punchItemDetails = PunchItemDetailsDtoMock(_raisedByOrg, _clearingByOrg);
+
+        _punchItemServiceMock = Substitute.For<IPunchItemService>();
+        _punchItemServiceMock.GetByCheckListGuid(_punchItemDetails.CheckListGuid, default)
+            .Returns(new List<PunchItemDetailsDto>(){ _punchItemDetails });
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs:38]: remove extra ()

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+        _query = new GetPunchItemsByCheckListGuidQuery(_punchItemDetails.CheckListGuid);
+        _dut = new GetPunchItemsByCheckListGuidQueryHandler(_punchItemServiceMock);
+    }
+
+    [TestMethod]
+    public async Task Handle_ShouldReturnCorrectCreatedPunchItems()
+    {
+        // Act
+        var result = await _dut.Handle(_query, default);
+        var data = result.Data?.ToList();
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/GetPunchItemsByCheckList/GetPunchItemsByCheckListGuidQueryHandlerTests.cs:49]: move to line 53

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/PunchItemService/PunchItemServiceTests.cs [hunk 29]
```diff
@@ -0,0 +1,688 @@
...
+    {
+        // Arrange
+        await using var context = new CompletionContext(_dbContextOptions, _plantProviderMock, _eventDispatcherMock, _currentUserProviderMock, _tokenCredentialsMock);
+        var dut = new PunchItemServices.PunchItemService(context);
+
+        // Act
+        var result = await dut.GetByCheckListGuid(_checkListGuid, default);
+
+        // Assert
+        Assert.IsTrue(2 == result.Count(x => x.CheckListGuid == _checkListGuid));
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/PunchItemService/PunchItemServiceTests.cs:142]: image

File: src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs [hunk 30]
```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.Query/PunchItemServices/PunchItemService.cs:3]: remove

File: src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs [hunk 31]
```diff
@@ -116,6 +116,7 @@ public static void AddApplicationModules(this IServiceCollection services, IConf
         services.AddScoped<Command.Attachments.IAttachmentService, Command.Attachments.AttachmentService>();
         services.AddScoped<Query.Attachments.IAttachmentService, Query.Attachments.AttachmentService>();
         services.AddScoped<Command.ModifiedEvents.IModifiedEventService, Command.ModifiedEvents.ModifiedEventService>();
+        services.AddScoped<Query.PunchItemServices.IPunchItemService, Query.PunchItemServices.PunchItemService>();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.Completion.WebApi/DiModules/ApplicationModule.cs:119]: Add using Equinor.ProCoSys.Completion.Query.PunchItemServices and simplify this to services.AddScoped<IPunchItemService, PunchItemService>();

File: src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/PunchItemService/PunchItemServiceTests.cs [hunk 32]
```diff
@@ -0,0 +1,688 @@
...
+using Equinor.ProCoSys.Completion.Domain.AggregateModels.PunchItemAggregate;
+using Equinor.ProCoSys.Completion.Infrastructure;
+using Equinor.ProCoSys.Completion.Test.Common;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Equinor.ProCoSys.Completion.Domain.AggregateModels.SWCRAggregate;
+using Equinor.ProCoSys.Completion.Domain.AggregateModels.WorkOrderAggregate;
+using Equinor.ProCoSys.Completion.Query.PunchItemServices;
+
+namespace Equinor.ProCoSys.Completion.Query.Tests.PunchItemQueries.PunchItemService;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.Query.Tests/PunchItemQueries/PunchItemService/PunchItemServiceTests.cs:18]: rename folder and namespace to Equinor.ProCoSys.Completion.Query.Tests.PunchItemQueries.PunchItemService**s**
Then you can simplify all
`var dut = new PunchItemServices.PunchItemService(context);` to `var dut = new PunchItemService(context);`

File: src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/KnownTestData.cs [hunk 33]
```diff
@@ -15,6 +15,7 @@ public KnownTestData(string plant)
 
     public string Plant { get; }
     public PunchItem PunchItem { get; set; }
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.Completion.WebApi.IntegrationTests/KnownTestData.cs:17]: rename to PunchItemA

**Review by janid1967 — APPROVED** [review]
Approved after minor fixes
